Update last inserted block state to track multiple blocks

getdave opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Companion/sub PR to https://github.com/WordPress/gutenberg/pull/46857.

Currently the last inserted block is tracked in state as a _single_ clientId. However this only accounts for the scenario where a _single_ block has been inserted.

This PR updates and addresses this concern by changing the state to track all inserted blocks and also providing new selectors to access this new state shape (whilst deprecating existing ones).


## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Currently the state only handle single blocks. It does not account for situations such as:

- `REPLACE_` type actions which may still "insert" new blocks but may provide multiple blocks.
- `INSERT_BLOCKS` action which may also provide multiple blocks

This is important as [sometimes](https://github.com/WordPress/gutenberg/pull/46857) it's important to know when `REPLACE_` type actions have caused new blocks to be inserted.

By altering the reducer state we can track this whilst still preserving the behaviour of the original selectors.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This PR:

- updates state reducer to handle tracking of _multiple_ inserted blocks.
- deprecates existing `getLastInsertedBlockClientId` selector which returned a single block only and is thus no longer suitable for accessing the state in an non-arbitrary way.
- adds a new selector `getLastInsertedBlocksClientIds` which returns an array of block clientIds.
- adds tests.
- include `REPLACE_` actions in the reducer to ensure these are also tracked as potential "insertions" of blocks

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

Tests:
- `npm run test:unit packages/block-editor/src/store/test/selectors.js`
- `npm run test:unit packages/block-editor/src/store/test/reducer.js`

You can also add some blocks in a new Post and then run:

```js
wp.data.select('core/block-editor').getBlocksByClientId( wp.data.select('core/block-editor').getLastInsertedBlocksClientIds() )
```

...and check that you see an array of blocks.

Also try adding some blocks, transforming into a quote and then unwrapping the quote back to individual blocks. Then run the selector above. You should see multiple blocks returned.

#### Check Depreciation Notice

Run 

```
wp.data.select('core/block-editor').getLastInsertedBlockClientId()
```

Check that the deprecation notice in console is correct.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

getdave committed:
Include replace in tracking last “inserted” block

getdave committed:
Update reducer tests

getdave committed:
Make selector experimental

getdave committed:
Refactor state to return array of inserted blocks

getdave commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2647,7 +2647,7 @@ export const __experimentalGetActiveBlockIdByBlockNames = createSelector(
 	const { lastBlockInserted } = state;
 	return (
-		lastBlockInserted.clientId === clientId &&
+		lastBlockInserted.clientIds?.includes( clientId ) &&
```
The idea here is to preserve the intent of the existing selector as if the block has just been inserted then it will stil be a valid result.

getdave commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2658,8 +2658,11 @@ export function wasBlockJustInserted( state, clientId, source ) {
+export function __experimentalGetLastInsertedBlockClientId( state ) {
+	return (
+		state?.lastBlockInserted?.clientIds?.length &&
+		state?.lastBlockInserted?.clientIds[ 0 ]
```
We _could_ update this selector to return _all_ the clientIds and allow the user to deal with which one they want to count as the "last" or "latest".

So if there is a single block you'd then consume it thus

```
const lastInsertedBlockClientId = __experimentalGetLastInsertedBlockClientIds()[0]
```

scruffian replied:
I think we should return the array.

youknowriad replied:
Agreed, picking the first one is very arbitrary

getdave replied:
Excellent. We have a consensus. So I assume we:

- rename the selector to `__experimentalGetLastInsertedBlockClientIds` (plural)
- deprecate the original selector

getdave replied:
I'm going to follow the guidance and advice in the docs

https://github.com/WordPress/gutenberg/blob/trunk/docs/contributors/code/backward-compatibility.md#how-to-preserve-backward-compatibility-for-a-javascript-function

getdave replied:
Ok I have now:

- deprecated the original selector (this is why it should have been experimental 🤦 )
- add a new selector which returns an array

I still don't know if the new selector should be experimental. @draganescu felt there was no need as it was in use by the offcanvas. But that itself is an experiment so ... 🤷‍♂️  Opinions appreciated.

scruffian replied:
I vote we keep it experimental for now.

draganescu commented:
I think these updates make sense. I don't think the selector should be experimental since it will be used by the Link UI.

draganescu reviewed (approved):
I think the updates here make sense. The `lastBlockInserted` reducer is surprising now by selecting the "first" from a list of inserted blocks which could be very well not the last inserted.

My one nit is `getLastInsertedBlockClientId` should **not** be experimental because it is very simple and it will be used by Link UI right away.

getdave commented:
> The `lastBlockInserted` reducer is surprising now by selecting the "first" from a list of inserted blocks which could be very well not the last inserted.

I think you mean the _selector_? The _reducer_ now tracks all blocks not just the first one.

I kept the selector "as is" but we could easily change it to return all blocks. What do you think @draganescu?

Commit:
rename selector

scruffian commented:
I pushed a fix to rename getLastInsertedBlockClientId to __experimentalGetLastInsertedBlockClientId as I was getting a JS error.

youknowriad commented on packages/block-editor/src/components/off-canvas-editor/block-contents.js:
```diff
@@ -51,12 +51,13 @@ const ListViewBlockContents = forwardRef(
 					selectedBlockInBlockEditor: getSelectedBlockClientId(),
-					lastInsertedBlockClientId: getLastInsertedBlockClientId(),
+					lastInsertedBlockClientId:
+						__experimentalGetLastInsertedBlockClientId(),
```
Did we merge the "@wordpress/experiments" PR yet, in which case prefixing is considered like an outdated approach?

youknowriad replied:
Maybe @adamziel would know the best approach to introduce an experimental selector right now?

getdave replied:
Yes that's a good point. It might be worth a separate PR to make it experimental and then rebase _this_ PR.

getdave replied:
Due to the flux in the API since introduced, I've followed up and made this a private selector 

https://github.com/WordPress/gutenberg/pull/47638

getdave committed:
Deprecates existing function

getdave committed:
Adds selector to return multiple blocks

getdave committed:
Add tests for new selector

getdave committed:
Fix usage to utilise new selector

getdave commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2647,19 +2647,43 @@ export const __experimentalGetActiveBlockIdByBlockNames = createSelector(
+	deprecated(
+		'wp.data.select( "core/block-editor" ).getLastInsertedBlockClientId',
+		{
+			since: '15.0',
```
Can I get a confidence check that this is the correct release?

scruffian replied:
14.9 was the last release so I believe so.

Commit:
fix selector name

scruffian reviewed (approved):
LGTM

youknowriad commented on docs/reference-guides/data/data-core-block-editor.md:
```diff
@@ -558,6 +558,8 @@ _Properties_
 
 ### getLastInsertedBlockClientId
```
If I'm not wrong this didn't land in WordPress at all, we should be able to remove it

getdave replied:
Ah so it doesn't count for Gutenberg Plugin just for WP Core? So as long as it didn't land in a WP release we're all good to remove?

getdave replied:
Removed in https://github.com/WordPress/gutenberg/pull/46885/commits/7fe30705446f6b4f6ee5a55ad57d9d8128147283

getdave committed:
Remove selector

As per @youknowriad advice in https://github.com/WordPress/gutenberg/pull/46885#discussion_r1062346033

youknowriad reviewed (approved):
If we're still uncertain about this selector, it would be cool to mark it experimental once the experiments package lands.

bph commented:
Added the "Needs Dev Note" label to put in our our radar for WordPress 6.2 Fieldguide
Three options:
- [ ] stand-alone Dev note post on make blog
- [ ]  Mentioned is "Miscellaneous Editor Updates" dev note post
- [ ]  Listed under "But Wait there is more" section of the Fieldguide.

... should it make it into this release.

adamziel commented:
@getdave to make this selector experimental with [the new lock()/unlock() API](https://github.com/WordPress/gutenberg/pull/46131) you'll need to:

1. Opt-in to experiments in the block-editor package – @noisysocks [has a PR where this is done](https://github.com/WordPress/gutenberg/pull/47229/files#diff-9a65e4c2c05471cae398d28f34f233b9d89d9718501218b3d87a4552270e75e7)
2. In `block-editor/store/index.js` registering the selector as private:

```js
// The `lock` function is exported from the internal experiments.js file where
// the opt-in function was called.
import { unlock } from './experiments';
import { getLastInsertedBlocksClientIds, ...selectors } from './selectors';

export const store = registerStore(/* ... */);
// Attach a private action to the exported store:
unlock( store ).registerPrivateSelector({
	getLastInsertedBlocksClientIds
} );
```
3. In `block-contents.js` use the private selector:

```js
import { unlock } from '../../experiments';

const {
	hasBlockMovingClientId,
	getSelectedBlockClientId,
	getLastInsertedBlockClientId,
	getLastInsertedBlocksClientIds,
} = unlock( select( blockEditorStore ) );
```

getdave commented:
This selector will be private https://github.com/WordPress/gutenberg/pull/47638
